src/Symfony/Component/DependencyInjection/Loader/FileLoader.php

```diff
@@ -34,4 +39,116 @@ public function __construct(ContainerBuilder $container, FileLocatorInterface $l
+    }
+
+    public function registerClasses(Definition $prototype, $psr4Glob, $directory)
+    {
+        $directory = $this->locator->locate($directory, $this->currentDir, true);
+        $directory = realpath($directory) ?: $directory;
+        if (!$classes = $this->findClasses($psr4Glob, $directory)) {
+            $this->container->getCompiler()->addLogMessage(sprintf('No classes matching "%s" were found in directory "%s".', $psr4Glob, $directory));
+        }
+
+        foreach ($classes as $class) {
+            $this->container->setDefinition($class, clone $prototype);
+        }
+    }
+
+    private function findClasses($psr4Glob, $directory)
+    {
+        $missingClassException = new \ReflectionException();
+        $throwingAutoloader = function () use ($missingClassException) { throw $missingClassException; };
+        spl_autoload_register($throwingAutoloader);
+        $classes = array();
+
+        try {
+            if (!preg_match('/^(?:[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+\\\\)++/', $psr4Glob, $match)) {
+                throw new InvalidArgumentException(sprintf('Namespace glob does not start with a valid PSR-4 prefix: %s.', $psr4Glob));
+            }
+
+            $psr4Prefix = $match[0];
+            $classRegexp = $this->patternToRegexp($psr4Glob);
+            $excludeTest = false === strpos($psr4Glob, 'Test', strlen($psr4Prefix));
+
+            $flags = \FilesystemIterator::SKIP_DOTS | \FilesystemIterator::UNIX_PATHS | \FilesystemIterator::FOLLOW_SYMLINKS;
+            $finder = new \RecursiveDirectoryIterator($directory, $flags);
+            $extRegexp = defined('HHVM_VERSION') ? '/\\.(?:php|hh)$/' : '/\\.php$/';
+            $prefixLen = 1 + strlen($directory);
+
+            // track directories only for new & removed files
+            $this->container->addResource(new DirectoryResource(realpath($directory) ?: $directory, '/^$/'));
+
+            foreach (new \RecursiveIteratorIterator($finder) as $path => $info) {
+                if (preg_match($extRegexp, $path, $m) && $info->isFile() && $info->isReadable()) {
+                    $class = str_replace('/', '\\', substr($path, $prefixLen, -strlen($m[0])));
+
+                    if ($excludeTest && false !== strpos($class, 'Test')) {
+                        continue;
+                    }
+                    if (!preg_match($classRegexp, $class = $psr4Prefix.$class)) {
+                        continue;
+                    }
+                    if (!preg_match('/^[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+(?:\\\\[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+)*+$/', $class)) {
+                        continue;
+                    }
+                    if (class_exists($class, false)) {
+                        $classes[] = $class;
+                        continue;
+                    }
+                    if (!function_exists('opcache_is_script_cached') || !opcache_is_script_cached($info->getRealPath() ?: $path)) {
```

Comment:
wouldn't something rejected by this check, be also rejected by the `class_exists` below?